JmsMessagingTemplate.convertSendAndReceive fails re-declaring temp queue

Using springboot 1.5.11.RELEASE, org.springframework:spring-jms:4.3.15, com.rabbitmq.jms:rabbitmq-jms:1.8.1

	@Autowired
	JmsMessagingTemplate t;

	@GetMapping("/test")
	String test() {
		String resp = t.convertSendAndReceive("tq", "request", String.class);
		return resp;
	}

	@JmsListener(destination = "tq")
	String list(Message msg) {
		System.out.println("Received message" + msg);
		return "response";
	}


Got below exception

Received messagecom.rabbitmq.jms.client.message.RMQTextMessage@35e0782
2018-04-09 13:42:10.604 ERROR 6340 --- [enerContainer-1] com.rabbitmq.jms.client.RMQSession       : RabbitMQ exception on queue declare name(jms-temp-queue-dc5833f7-1870-4a20-b60e-cfc2b93c8dd7), durable(false), exclusive(true), auto-delete(false), properties(null)

```
java.io.IOException: null
	at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:124) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:120) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:142) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.ChannelN.queueDeclare(ChannelN.java:952) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.queueDeclare(AutorecoveringChannel.java:333) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.jms.client.RMQSession.declareRMQQueue(RMQSession.java:838) [rabbitmq-jms-1.8.1.jar:1.8.1]
	at com.rabbitmq.jms.client.RMQSession.declareDestinationIfNecessary(RMQSession.java:615) [rabbitmq-jms-1.8.1.jar:1.8.1]
	at com.rabbitmq.jms.client.RMQSession.createProducer(RMQSession.java:606) [rabbitmq-jms-1.8.1.jar:1.8.1]
	at org.springframework.jms.listener.adapter.AbstractAdaptableMessageListener.sendResponse(AbstractAdaptableMessageListener.java:403) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.adapter.AbstractAdaptableMessageListener.handleResult(AbstractAdaptableMessageListener.java:253) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:71) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:719) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:679) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:649) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.doReceiveAndExecute(AbstractPollingMessageListenerContainer.java:317) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.receiveAndExecute(AbstractPollingMessageListenerContainer.java:255) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.invokeListener(DefaultMessageListenerContainer.java:1168) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.executeOngoingLoop(DefaultMessageListenerContainer.java:1160) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.run(DefaultMessageListenerContainer.java:1057) [spring-jms-4.3.15.RELEASE.jar:4.3.15.RELEASE]
	at java.lang.Thread.run(Unknown Source) [na:1.8.0_162]
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; protocol method: #method<channel.close>(reply-code=405, reply-text=RESOURCE_LOCKED - cannot obtain exclusive access to locked queue 'jms-temp-queue-dc5833f7-1870-4a20-b60e-cfc2b93c8dd7' in vhost '/', class-id=50, method-id=10)
	at com.rabbitmq.utility.ValueOrException.getValue(ValueOrException.java:66) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:36) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:443) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQChannel.privateRpc(AMQChannel.java:263) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:136) ~[amqp-client-4.3.0.jar:4.3.0]
	... 17 common frames omitted
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; protocol method: #method<channel.close>(reply-code=405, reply-text=RESOURCE_LOCKED - cannot obtain exclusive access to locked queue 'jms-temp-queue-dc5833f7-1870-4a20-b60e-cfc2b93c8dd7' in vhost '/', class-id=50, method-id=10)
	at com.rabbitmq.client.impl.ChannelN.asyncShutdown(ChannelN.java:509) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.ChannelN.processAsync(ChannelN.java:340) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQChannel.handleCompleteInboundCommand(AMQChannel.java:162) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQChannel.handleFrame(AMQChannel.java:109) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQConnection.readFrame(AMQConnection.java:643) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQConnection.access$300(AMQConnection.java:47) ~[amqp-client-4.3.0.jar:4.3.0]
	at com.rabbitmq.client.impl.AMQConnection$MainLoop.run(AMQConnection.java:581) ~[amqp-client-4.3.0.jar:4.3.0]
	... 1 common frames omitted
```

--- a/src/main/java/com/rabbitmq/jms/admin/RMQConnectionFactory.java
+++ b/src/main/java/com/rabbitmq/jms/admin/RMQConnectionFactory.java
@@ -7,6 +7,7 @@ import com.rabbitmq.client.MetricsCollector;
 import com.rabbitmq.client.NoOpMetricsCollector;
 import com.rabbitmq.jms.client.ConnectionParams;
 import com.rabbitmq.jms.client.RMQConnection;
+import com.rabbitmq.jms.client.SendingContextConsumer;
 import com.rabbitmq.jms.util.RMQJMSException;
 import com.rabbitmq.jms.util.RMQJMSSecurityException;
 import com.rabbitmq.jms.util.WhiteListObjectInputStream;
@@ -35,7 +36,6 @@ import java.io.Serializable;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.security.NoSuchAlgorithmException;
-import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.TimeoutException;
@@ -93,12 +93,14 @@ public class RMQConnectionFactory implements ConnectionFactory, Referenceable, S
      * If set to true, those queues will be deleted when the session is closed.
      * If set to false, queues will be deleted when the owning connection is closed.
      * Default is false.
+     *
      * @since 1.8.0
      */
     private boolean cleanUpServerNamedQueuesForNonDurableTopicsOnSessionClose = false;
 
     /**
      * Callback to customise properties of outbound AMQP messages.
+     *
      * @since 1.9.0
      */
     private BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser;
@@ -111,11 +113,18 @@ public class RMQConnectionFactory implements ConnectionFactory, Referenceable, S
     private MetricsCollector metricsCollector = new NoOpMetricsCollector();
 
     /**
-     * For post-processor the {@link com.rabbitmq.client.ConnectionFactory} before creating the AMQP connection.
+     * For post-processing the {@link com.rabbitmq.client.ConnectionFactory} before creating the AMQP connection.
      *
      * @since 1.10.0
      */
     private Consumer<com.rabbitmq.client.ConnectionFactory> amqpConnectionFactoryPostProcessor = cf -> {};
+
+    /**
+     * Callback before sending a message.
+     *
+     * @since 1.11.0
+     */
+    private SendingContextConsumer sendingContextConsumer = ctx -> {};
 
     /** Default not to use ssl */
     private boolean ssl = false;
@@ -223,6 +232,7 @@ public class RMQConnectionFactory implements ConnectionFactory, Referenceable, S
             .setRequeueOnMessageListenerException(requeueOnMessageListenerException)
             .setCleanUpServerNamedQueuesForNonDurableTopicsOnSessionClose(this.cleanUpServerNamedQueuesForNonDurableTopicsOnSessionClose)
             .setAmqpPropertiesCustomiser(amqpPropertiesCustomiser)
+            .setSendingContextConsumer(sendingContextConsumer)
         );
         conn.setTrustedPackages(this.trustedPackages);
         logger.debug("Connection {} created.", conn);
@@ -825,6 +835,19 @@ public class RMQConnectionFactory implements ConnectionFactory, Referenceable, S
      */
     public void setAmqpConnectionFactoryPostProcessor(Consumer<com.rabbitmq.client.ConnectionFactory> amqpConnectionFactoryPostProcessor) {
         this.amqpConnectionFactoryPostProcessor = amqpConnectionFactoryPostProcessor;
+    }
+
+    /**
+     * Set callback called before sending a message.
+     * Can be used to customize the message or the destination
+     * before the message is actually sent.
+     *
+     * @see SendingContextConsumer
+     * @see com.rabbitmq.jms.client.SendingContext
+     * @since 1.11.0
+     */
+    public void setSendingContextConsumer(SendingContextConsumer sendingContextConsumer) {
+        this.sendingContextConsumer = sendingContextConsumer;
     }
 
     @FunctionalInterface

--- a/src/main/java/com/rabbitmq/jms/client/ConnectionParams.java
+++ b/src/main/java/com/rabbitmq/jms/client/ConnectionParams.java
@@ -52,15 +52,24 @@ public class ConnectionParams {
      * If set to true, those queues will be deleted when the session is closed.
      * If set to false, queues will be deleted when the owning connection is closed.
      * Default is false.
+     *
      * @since 1.8.0
      */
     private boolean cleanUpServerNamedQueuesForNonDurableTopicsOnSessionClose = false;
 
     /**
      * Callback to customise properties of outbound AMQP messages.
+     *
      * @since 1.9.0
      */
     private BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser;
+
+    /**
+     * Callback before sending a message.
+     *
+     * @since 1.11.0
+     */
+    private SendingContextConsumer sendingContextConsumer;
 
     public Connection getRabbitConnection() {
         return rabbitConnection;
@@ -140,6 +149,15 @@ public class ConnectionParams {
 
     public ConnectionParams setAmqpPropertiesCustomiser(BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser) {
         this.amqpPropertiesCustomiser = amqpPropertiesCustomiser;
+        return this;
+    }
+
+    public SendingContextConsumer getSendingContextConsumer() {
+        return sendingContextConsumer;
+    }
+
+    public ConnectionParams setSendingContextConsumer(SendingContextConsumer sendingContextConsumer) {
+        this.sendingContextConsumer = sendingContextConsumer;
         return this;
     }
 }

--- a/src/main/java/com/rabbitmq/jms/client/RMQConnection.java
+++ b/src/main/java/com/rabbitmq/jms/client/RMQConnection.java
@@ -104,9 +104,17 @@ public class RMQConnection implements Connection, QueueConnection, TopicConnecti
 
     /**
      * Callback to customise properties of outbound AMQP messages.
+     *
      * @since 1.9.0
      */
     private final BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser;
+
+    /**
+     * Callback before sending a message.
+     *
+     * @since 1.11.0
+     */
+    private final SendingContextConsumer sendingContextConsumer;
 
     /**
      * Classes in these packages can be transferred via ObjectMessage.
@@ -132,6 +140,7 @@ public class RMQConnection implements Connection, QueueConnection, TopicConnecti
         this.requeueOnMessageListenerException = connectionParams.willRequeueOnMessageListenerException();
         this.cleanUpServerNamedQueuesForNonDurableTopicsOnSessionClose = connectionParams.isCleanUpServerNamedQueuesForNonDurableTopicsOnSessionClose();
         this.amqpPropertiesCustomiser = connectionParams.getAmqpPropertiesCustomiser();
+        this.sendingContextConsumer = connectionParams.getSendingContextConsumer();
     }
 
     /**
@@ -181,6 +190,7 @@ public class RMQConnection implements Connection, QueueConnection, TopicConnecti
             .setRequeueOnMessageListenerException(this.requeueOnMessageListenerException)
             .setCleanUpServerNamedQueuesForNonDurableTopics(this.cleanUpServerNamedQueuesForNonDurableTopicsOnSessionClose)
             .setAmqpPropertiesCustomiser(this.amqpPropertiesCustomiser)
+            .setSendingContextConsumer(this.sendingContextConsumer)
         );
         session.setTrustedPackages(this.trustedPackages);
         this.sessions.add(session);

--- a/src/main/java/com/rabbitmq/jms/client/RMQMessageProducer.java
+++ b/src/main/java/com/rabbitmq/jms/client/RMQMessageProducer.java
@@ -72,8 +72,11 @@ public class RMQMessageProducer implements MessageProducer, QueueSender, TopicPu
 
     private final BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser;
 
+    private final SendingContextConsumer sendingContextConsumer;
+
     public RMQMessageProducer(RMQSession session, RMQDestination destination, boolean preferProducerMessageProperty,
-            BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser) {
+            BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser,
+            SendingContextConsumer sendingContextConsumer) {
         this.session = session;
         this.destination = destination;
         if (preferProducerMessageProperty) {
@@ -82,6 +85,12 @@ public class RMQMessageProducer implements MessageProducer, QueueSender, TopicPu
             sendingStrategy = new PreferMessagePropertySendingStrategy();
         }
         this.amqpPropertiesCustomiser = amqpPropertiesCustomiser == null ? (builder, message) -> builder : amqpPropertiesCustomiser;
+        this.sendingContextConsumer = sendingContextConsumer == null ? ctx -> {} : sendingContextConsumer;
+    }
+
+    public RMQMessageProducer(RMQSession session, RMQDestination destination, boolean preferProducerMessageProperty,
+        BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser) {
+        this(session, destination, preferProducerMessageProperty, amqpPropertiesCustomiser, ctx -> {});
     }
 
     /**
@@ -248,6 +257,8 @@ public class RMQMessageProducer implements MessageProducer, QueueSender, TopicPu
 
     private void internalSend(RMQDestination destination, Message message, int deliveryMode, int priority, long timeToLiveOrExpiration, MessageExpirationType messageExpirationType) throws JMSException {
         logger.trace("send/publish message({}) to destination({}) with properties deliveryMode({}), priority({}), timeToLive({})", message, destination, deliveryMode, priority, timeToLiveOrExpiration);
+
+        this.sendingContextConsumer.accept(new SendingContext(destination, message));
 
         if (destination == null)
             destination = this.destination;

--- a/src/main/java/com/rabbitmq/jms/client/RMQSession.java
+++ b/src/main/java/com/rabbitmq/jms/client/RMQSession.java
@@ -107,9 +107,17 @@ public class RMQSession implements Session, QueueSession, TopicSession {
 
     /**
      * Callback to customise properties of outbound AMQP messages.
+     *
      * @since 1.9.0
      */
     private final BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser;
+
+    /**
+     * Callback before sending a message.
+     *
+     * @since 1.11.0
+     */
+    private final SendingContextConsumer sendingContextConsumer;
 
     /** The main RabbitMQ channel we use under the hood */
     private final Channel channel;
@@ -204,6 +212,7 @@ public class RMQSession implements Session, QueueSession, TopicSession {
         this.requeueOnMessageListenerException = sessionParams.willRequeueOnMessageListenerException();
         this.cleanUpServerNamedQueuesForNonDurableTopics = sessionParams.isCleanUpServerNamedQueuesForNonDurableTopics();
         this.amqpPropertiesCustomiser = sessionParams.getAmqpPropertiesCustomiser();
+        this.sendingContextConsumer = sessionParams.getSendingContextConsumer();
 
         if (transacted) {
             this.acknowledgeMode = Session.SESSION_TRANSACTED;
@@ -613,7 +622,8 @@ public class RMQSession implements Session, QueueSession, TopicSession {
         illegalStateExceptionIfClosed();
         RMQDestination dest = (RMQDestination) destination;
         declareDestinationIfNecessary(dest);
-        RMQMessageProducer producer = new RMQMessageProducer(this, dest, this.preferProducerMessageProperty, this.amqpPropertiesCustomiser);
+        RMQMessageProducer producer = new RMQMessageProducer(this, dest, this.preferProducerMessageProperty,
+            this.amqpPropertiesCustomiser, this.sendingContextConsumer);
         this.producers.add(producer);
         return producer;
     }

--- a/src/main/java/com/rabbitmq/jms/client/SendingContext.java
+++ b/src/main/java/com/rabbitmq/jms/client/SendingContext.java
@@ -1,0 +1,33 @@
+/* Copyright (c) 2018 Pivotal Software, Inc. All rights reserved. */
+
+package com.rabbitmq.jms.client;
+
+import javax.jms.Destination;
+import javax.jms.Message;
+
+/**
+ * Context when sending message.
+ *
+ * @see com.rabbitmq.jms.admin.RMQConnectionFactory#setSendingContextConsumer(SendingContextConsumer)
+ * @see SendingContextConsumer
+ * @since 1.11.0
+ */
+public class SendingContext {
+
+    private final Message message;
+
+    private final Destination destination;
+
+    public SendingContext(Destination destination, Message message) {
+        this.destination = destination;
+        this.message = message;
+    }
+
+    public Destination getDestination() {
+        return destination;
+    }
+
+    public Message getMessage() {
+        return message;
+    }
+}

--- a/src/main/java/com/rabbitmq/jms/client/SendingContextConsumer.java
+++ b/src/main/java/com/rabbitmq/jms/client/SendingContextConsumer.java
@@ -1,0 +1,43 @@
+package com.rabbitmq.jms.client;
+
+import javax.jms.JMSException;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * Callback before sending a message.
+ *
+ * @see com.rabbitmq.jms.admin.RMQConnectionFactory#setSendingContextConsumer(SendingContextConsumer)
+ * @see SendingContext
+ * @since 1.11.0
+ */
+@FunctionalInterface
+public interface SendingContextConsumer {
+
+    /**
+     * Called before sending a message.
+     * <p>
+     * Can be used to customize the message or the destination
+     * before the message is actually sent.
+     *
+     * @param sendingContext
+     * @throws JMSException
+     */
+    void accept(SendingContext sendingContext) throws JMSException;
+
+    /**
+     * Same semantics as {@link Consumer#andThen(Consumer)}.
+     *
+     * @param after the operation to perform after this operation
+     * @return a composed {@code SendingContextConsumer} that performs in sequence this
+     * operation followed by the {@code after} operation
+     * @throws NullPointerException if {@code after} is null
+     */
+    default SendingContextConsumer andThen(SendingContextConsumer after) {
+        Objects.requireNonNull(after);
+        return (SendingContext ctx) -> {
+            accept(ctx);
+            after.accept(ctx);
+        };
+    }
+}

--- a/src/main/java/com/rabbitmq/jms/client/SessionParams.java
+++ b/src/main/java/com/rabbitmq/jms/client/SessionParams.java
@@ -48,15 +48,24 @@ public class SessionParams {
      * If set to true, those queues will be deleted when the session is closed.
      * If set to false, queues will be deleted when the owning connection is closed.
      * Default is false.
+     *
      * @since 1.8.0
      */
     private boolean cleanUpServerNamedQueuesForNonDurableTopics = false;
 
     /**
      * Callback to customise properties of outbound AMQP messages.
+     *
      * @since 1.9.0
      */
     private BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser;
+
+    /**
+     * Callback before sending a message.
+     *
+     * @since 1.11.0
+     */
+    private SendingContextConsumer sendingContextConsumer;
 
     public RMQConnection getConnection() {
         return connection;
@@ -137,5 +146,14 @@ public class SessionParams {
     public SessionParams setAmqpPropertiesCustomiser(BiFunction<AMQP.BasicProperties.Builder, Message, AMQP.BasicProperties.Builder> amqpPropertiesCustomiser) {
         this.amqpPropertiesCustomiser = amqpPropertiesCustomiser;
         return this;
+    }
+
+    public SessionParams setSendingContextConsumer(SendingContextConsumer sendingContextConsumer) {
+        this.sendingContextConsumer = sendingContextConsumer;
+        return this;
+    }
+
+    public SendingContextConsumer getSendingContextConsumer() {
+        return sendingContextConsumer;
     }
 }

--- a/src/test/java/com/rabbitmq/integration/tests/RpcIT.java
+++ b/src/test/java/com/rabbitmq/integration/tests/RpcIT.java
@@ -1,0 +1,173 @@
+/* Copyright (c) 2018 Pivotal Software, Inc. All rights reserved. */
+package com.rabbitmq.integration.tests;
+
+import com.rabbitmq.jms.admin.RMQConnectionFactory;
+import com.rabbitmq.jms.admin.RMQDestination;
+import com.rabbitmq.jms.client.SendingContextConsumer;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.Queue;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+import java.util.UUID;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+/**
+ *
+ */
+public class RpcIT {
+
+    private static final String QUEUE_NAME = "test.queue." + RpcIT.class.getCanonicalName();
+    private static final Logger LOGGER = LoggerFactory.getLogger(RpcIT.class);
+
+    Connection serverConnection, clientConnection;
+
+    RpcServer rpcServer;
+
+    protected static void drainQueue(Session session, Queue queue) throws Exception {
+        MessageConsumer receiver = session.createConsumer(queue);
+        Message msg = receiver.receiveNoWait();
+        while (msg != null) {
+            msg = receiver.receiveNoWait();
+        }
+    }
+
+    static SendingContextConsumer destinationAlreadyDeclaredForRpcResponse() {
+        return ctx -> {
+            if (ctx.getMessage().getJMSCorrelationID() != null && ctx.getDestination() instanceof RMQDestination) {
+                RMQDestination destination = (RMQDestination) ctx.getDestination();
+                destination.setDeclared(true);
+            }
+        };
+    }
+
+    @BeforeEach
+    public void init() throws Exception {
+        ConnectionFactory connectionFactory = AbstractTestConnectionFactory.getTestConnectionFactory()
+            .getConnectionFactory();
+        clientConnection = connectionFactory.createConnection();
+        clientConnection.start();
+    }
+
+    @AfterEach
+    public void tearDown() throws Exception {
+        rpcServer.close();
+        if (clientConnection != null) {
+            clientConnection.close();
+        }
+        if (serverConnection != null) {
+            serverConnection.close();
+        }
+        com.rabbitmq.client.ConnectionFactory cf = new com.rabbitmq.client.ConnectionFactory();
+        try (com.rabbitmq.client.Connection c = cf.newConnection()) {
+            c.createChannel().queueDelete(QUEUE_NAME);
+        }
+    }
+
+    @Test
+    public void noResponseWhenServerTriesToRecreateTemporaryResponseQueue() throws Exception {
+        setupRpcServer();
+
+        String messageContent = UUID.randomUUID().toString();
+        Message response = doRpc(messageContent);
+        assertNull(response);
+    }
+
+    @Test
+    public void responseOkWhenServerDoesNotRecreateTemporaryResponseQueue() throws Exception {
+        setupRpcServer(destinationAlreadyDeclaredForRpcResponse());
+
+        String messageContent = UUID.randomUUID().toString();
+        Message response = doRpc(messageContent);
+        assertNotNull(response);
+        assertThat(response, is(instanceOf(TextMessage.class)));
+        assertThat(((TextMessage) response).getText(), is("*** " + messageContent + " ***"));
+    }
+
+    Message doRpc(String messageContent) throws Exception {
+        Session session = clientConnection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        TextMessage message = session.createTextMessage(messageContent);
+        message.setJMSCorrelationID(messageContent);
+
+        Destination replyQueue = session.createTemporaryQueue();
+        MessageProducer producer = session.createProducer(session.createQueue(QUEUE_NAME));
+
+        MessageConsumer responseConsumer = session.createConsumer(replyQueue);
+        BlockingQueue<Message> queue = new ArrayBlockingQueue<>(1);
+        responseConsumer.setMessageListener(msg -> queue.add(msg));
+        message.setJMSReplyTo(replyQueue);
+        producer.send(message);
+        return queue.poll(2, TimeUnit.SECONDS);
+    }
+
+    void setupRpcServer(SendingContextConsumer sendingContextConsumer) throws Exception {
+        RMQConnectionFactory connectionFactory = (RMQConnectionFactory) AbstractTestConnectionFactory.getTestConnectionFactory()
+            .getConnectionFactory();
+        connectionFactory.setSendingContextConsumer(sendingContextConsumer);
+        serverConnection = connectionFactory.createConnection();
+        serverConnection.start();
+        Session session = serverConnection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        Queue queue = session.createQueue(QUEUE_NAME);
+        drainQueue(session, queue);
+        session.close();
+        rpcServer = new RpcServer(serverConnection);
+    }
+
+    void setupRpcServer() throws Exception {
+        setupRpcServer(ctx -> {
+        });
+    }
+
+    private static class RpcServer {
+
+        Session session;
+
+        public RpcServer(Connection connection) throws JMSException {
+            session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+            Destination destination = session.createQueue(QUEUE_NAME);
+            MessageProducer replyProducer = session.createProducer(null);
+            MessageConsumer consumer = session.createConsumer(destination);
+            consumer.setMessageListener(msg -> {
+                TextMessage message = (TextMessage) msg;
+                try {
+                    String text = message.getText();
+                    Destination replyQueue = message.getJMSReplyTo();
+                    if (replyQueue != null) {
+                        TextMessage replyMessage = session.createTextMessage("*** " + text + " ***");
+                        replyMessage.setJMSCorrelationID(message.getJMSCorrelationID());
+                        replyProducer.send(replyQueue, replyMessage);
+                    }
+                } catch (JMSException e) {
+                    LOGGER.warn("Error in RPC server", e);
+                }
+            });
+        }
+
+        void close() {
+            try {
+                session.close();
+            } catch (Exception e) {
+
+            }
+        }
+    }
+}

